### PR TITLE
HS-632: Implement datachoices as a new service (1/2)

### PR DESCRIPTION
**Pull request by thomasbigger584**
## Description
HS-632: Implement datachoices as a new service (1/2)

Adding quartz instead of an in-memory scheduler which was previously used, to help with scalability, sychronization and resilience of the scheduler

Supporting tenant id so multiple tenants can opt in and we can perform queries based on opt-ed in tenants.

Will follow up with another PR with more implementation as this one is getting too large.

## Jira link(s)
- https://issues.opennms.org/browse/HS-632

## Checklist
* [x] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.

### CODE REVIEW

File: shared-lib/datachoices/src/main/java/org/opennms/horizon/datachoices/Constants.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.horizon.datachoices;
+
+import io.grpc.Context;
+import io.grpc.Metadata;
+
+public interface Constants {
```
Comment (yangli2014) [inline shared-lib/datachoices/src/main/java/org/opennms/horizon/datachoices/Constants.java:34]: I have a PR to create global-wide GrpcConstants. https://github.com/OpenNMS/horizon-stream/pull/547

Comment (thomasbigger584) [reply shared-lib/datachoices/src/main/java/org/opennms/horizon/datachoices/Constants.java:34]: Sure will take your code when it merges

File: .github/actions/datachoices/action.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
+name: 'Build and test inventory'
```
Comment (yangli2014) [inline .github/actions/datachoices/action.yml]: Wrong name

Comment (thomasbigger584) [reply .github/actions/datachoices/action.yml]: Updated in recent commit

File: datachoices/src/main/java/org/opennms/horizon/datachoices/grpc/DataChoicesGrpcService.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    public void toggle(ToggleDataChoicesDTO request, StreamObserver<ToggleDataChoicesDTO> responseObserver) {
+        Optional<String> tenantIdOptional = tenantLookup.lookupTenantId(Context.current());
+
+        if (tenantIdOptional.isEmpty()) {
+            Status status = Status.newBuilder()
+                .setCode(Code.INVALID_ARGUMENT_VALUE)
+                .setMessage("Tenant Id can't be empty")
+                .build();
+            responseObserver.onError(StatusProto.toStatusRuntimeException(status));
+            return;
```
Comment (yangli2014) [inline datachoices/src/main/java/org/opennms/horizon/datachoices/grpc/DataChoicesGrpcService.java:62]: Don't need to return here.

File: datachoices/src/main/java/org/opennms/horizon/datachoices/grpc/DataChoicesGrpcService.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        String tenantId = tenantIdOptional.get();
+
+        try {
+            service.toggle(request, tenantId);
+
+            responseObserver.onNext(request);
+            responseObserver.onCompleted();
+        } catch (Exception e) {
+            responseObserver.onError(e);
+        }
```
Comment (yangli2014) [inline datachoices/src/main/java/org/opennms/horizon/datachoices/grpc/DataChoicesGrpcService.java:74]:
```suggestion
tenantIdOptional.ifPresentOrElse(tenantId -> {
            service.toggle(request, tenantId);
            responseObserver.onNext(request);
            responseObserver.onCompleted();
        }, ()->{
            Status status = Status.newBuilder()
                .setCode(Code.INVALID_ARGUMENT_VALUE)
                .setMessage("Tenant Id can't be empty")
                .build();
            responseObserver.onError(StatusProto.toStatusRuntimeException(status));
        });
```

File: operator/api/v1alpha1/opennms_types.go [hunk 5]
```diff
@@ -15,170 +15,173 @@ limitations under the License.
 package v1alpha1
```
Comment (thomasbigger584) [inline operator/api/v1alpha1/opennms_types.go:15]: will revert this auto formatting

**Review by yangli2014 — APPROVED** [review]
LGTM

**Review by cgorantla — APPROVED** [review]
LGTM!

**Pull request merged**